### PR TITLE
refine kcnr to support enhanced topology status at each level

### PR DESCRIPTION
**Pull request by waynepeking348**
refine the structure for cnr to support a more flexible topology status

### CODE REVIEW

File: pkg/apis/node/v1alpha1/types.go [hunk 1]
```diff
@@ -92,13 +92,17 @@ type Property struct {
 
 type CustomNodeResourceStatus struct {
```
Comment (caohe) [inline pkg/apis/node/v1alpha1/types.go:93]: Do we need to add some fields to indicate the policy and the scope of topology manager?

**Review by csfldf — DISMISSED** [review]

File: pkg/apis/node/v1alpha1/types.go (hunk 1 above)
Comment (waynepeking348) [reply pkg/apis/node/v1alpha1/types.go:93]: the policy is defined as FeatureGate in Kubelet, but won't be defined in CNR for now; the reason for this requirement is for scheduler to understand, right?

Comment (caohe) [reply pkg/apis/node/v1alpha1/types.go:93]: Yes, these states need to be reported to the scheduler to decide which strategy to use.

Comment (waynepeking348) [reply pkg/apis/node/v1alpha1/types.go:93]: make sense, i will add a new field here to indicate this specific info

File: pkg/apis/node/v1alpha1/types.go [hunk 2]
```diff
@@ -92,19 +92,35 @@ type Property struct {
...
+	Resources Resources `json:"resources"`
+
 	// +optional
-	TopologyStatus *TopologyStatus `json:"topologyStatus,omitempty"`
+	TopologyStatus []*TopologyStatus `json:"topologyStatus,omitempty"`
+
+	// TopologyPolicy indicates placement policy for scheduler or other centralized components to follow.
+	// this policy (including topology scope) is defined in topology-manager, katalyst is
+	// responsible to parse the policy, and transform to TopologyPolicy here.
+	// +kubebuilder:default:=TopologyPolicyNode
```
Comment (caohe) [inline pkg/apis/node/v1alpha1/types.go:110]: maybe there is a typo: TopologyPolicyNode -> TopologyPolicyNone

**Review by caohe — CHANGES_REQUESTED** [review]
maybe there is a typo.

File: pkg/apis/node/v1alpha1/types.go (hunk 2 above)
Comment (caohe) [reply pkg/apis/node/v1alpha1/types.go:110]: or should we set default:=none ?

Comment (waynepeking348) [reply pkg/apis/node/v1alpha1/types.go]: fixed

**Review by caohe — DISMISSED** [review]
lgtm
Perhaps we should adapt katalyst-scheduler and reporter to these changes in KCNR API before merging this pr?

File: pkg/apis/node/v1alpha1/types.go [hunk 3]
```diff
@@ -92,19 +92,35 @@ type Property struct {
...
 type CustomNodeResourceStatus struct {
 	// +optional
+	// +kubebuilder:deprecatedversion:warning="this field is deprecated and should be moved to Resources"
 	ResourceAllocatable *v1.ResourceList `json:"resourceAllocatable,omitempty"`
 
 	// +optional
+	// +kubebuilder:deprecatedversion:warning="this field is deprecated and should be moved to Resources"
 	ResourceCapacity *v1.ResourceList `json:"resourceCapacity,omitempty"`
 
+	Resources Resources `json:"resources"`
```
Comment (caohe) [inline pkg/apis/node/v1alpha1/types.go:102]: Given that we already have `ResourceAllocatable` and `ResourceCapacity`, do we need the field `Resources`?

File: pkg/apis/node/v1alpha1/types.go [hunk 4]
```diff
@@ -130,24 +146,61 @@ const (
...
+
+	Resources Resources `json:"resources"`
 
-type SocketStatus struct {
-	SocketID int `json:"socketID"`
 	// +optional
-	Numas []*NumaStatus `json:"numas,omitempty"`
+	// +patchMergeKey=name
+	// +patchStrategy=merge
+	Attrs []Attribute `json:"attrs,omitempty" patchStrategy:"merge" patchMergeKey:"name"`
```
Comment (caohe) [inline pkg/apis/node/v1alpha1/types.go:164]: Perhaps we could retain the full spelling of this field as `Attributes`?

**Review by caohe — APPROVED** [review]
LGTM

File: pkg/apis/node/v1alpha1/types.go (hunk 3 above)
Comment (caohe) [reply pkg/apis/node/v1alpha1/types.go:95]: Maybe we need to add an omitempty tag to this field?

**Review by luomingmeng — APPROVED** [review]

**Review by csfldf — APPROVED** [review]

**Pull request merged**